### PR TITLE
[BugFix] fix expression in projection not been dict rewrited

### PR DESCRIPTION
**Pull request by packy92**
Signed-off-by: packy <wangchao@starrocks.com>

## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

For sql like:
```
select cast(max(s_address) as date) from supplier
```
Error happens when build fragment.
The cause is in `AddDecodeNodeForDictStringRule.rewriteOneScalarOperatorForProjection()`. After child been optimized, we may update the` stringColumnIdToDictColumnIds` in `DecodeContext` while the `allStringColumnIds` represent all low cardinality string cols in tables are immutable. If we use `allStringColumnIds` to check this expression whether can optimize, it may lead to a wrong result. 
When the child has encoded the cols, but we don't rewrite the usedCols in expression to new dict col, the error happens.

```
 if (!Projection.couldApplyDictOptimize(valueOperator, context.allStringColumnIds)) {
                return;
 }
```

BTW, I update the logic when evaluate whether we need add a decodeOp before projection. The new check logic is as noted in the comments.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java [hunk 1]
```diff
@@ -1192,12 +1192,14 @@ public void testMultiMaxMin() throws Exception {
                 "min(upper(S_COMMENT)) from supplier_nullable " +
                 "group by upper(S_COMMENT)";
         plan = getFragmentPlan(sql);
-        Assert.assertTrue(plan.contains("  5:Decode\n" +
+        Assert.assertTrue(plan, plan.contains("6:Decode\n" +
```
Comment (stdpain) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java:1195]: seems not as expected?

Comment (packy92) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java:1195]: The old
```
  6:Project
  |  <slot 11> : 11: max
  |  <slot 12> : 12
  |  
  5:Decode
  |  <dict id 17> : <string id 12>
  |  <dict id 15> : <string id 9>
  |  string functions:
  |  <function id 17> : DictExpr(14: S_COMMENT,[upper(<place-holder>)])
  |  <function id 15> : DictExpr(14: S_COMMENT,[upper(<place-holder>)])
  |  
  4:AGGREGATE (merge finalize)
  |  output: max(11: max), min(16: upper)
```

The new
```
6:Decode
  |  <dict id 17> : <string id 12>
  |  string functions:
  |  <function id 17> : DictExpr(14: S_COMMENT,[upper(<place-holder>)])
  |  
  5:Project
  |  <slot 11> : 11: max
  |  <slot 17> : 17: upper
  |  
  4:AGGREGATE (merge finalize)
  |  output: max(11: max), min(16: upper)
  |  group by: 15: upper
```
The new plan reduce a unnecessary col decode.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 35 / 36 (97.22%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|23|24|95.83%|[525]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/ColumnRefSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java [hunk 2]
```diff
@@ -211,30 +215,32 @@ private void visitProjectionBefore(OptExpression optExpression, DecodeContext co
...
             final Set<Integer> globalDictIds =
                     context.globalDicts.stream().map(a -> a.first).collect(Collectors.toSet());
 
-            // for each input in Projection,
-            // if input was dict column ,but we couldn't find it in global dict keys and input column
-            // is not generated by this project.
-            // (This means that the column is generated by the global dictionary through the expression)
-            // We insert a decode node to avoid not finding the input column
+            // for each entry in Projection:
+            // a. if it's a pass-through entry like col1 -> col1, we don't need to decode for it.
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java:228]: The last step of rewriting process will check if child has been encoded and fill the decode operator.
```
OptExpression rewriteExpr = root.getOp().accept(new DecodeVisitor(), root, context);
        if (context.hasEncoded) {
            return generateDecodeOExpr(context, Collections.singletonList(rewriteExpr));
        }
        return rewriteExpr;
```

**Review by kangkaisen — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java [hunk 3]
```diff
@@ -211,30 +215,32 @@ private void visitProjectionBefore(OptExpression optExpression, DecodeContext co
...
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
+                if (entry.getKey().equals(entry.getValue())) {
+                    continue;
+                }
+                final ColumnRefSet usedColumns = entry.getValue().getUsedColumns();
                 for (int cid : usedColumns.getColumnIds()) {
                     final Integer dictId = context.stringColumnIdToDictColumnIds.get(cid);
-                    if (dictId != null && !globalDictIds.contains(dictId) && dictColumnIds.contains(dictId) &&
-                            !projectOutputs.contains(dictId)) {
+                    if (dictId != null && !globalDictIds.contains(dictId)) {
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java:243]: Is it works for such plan?

Fragment2:
dict2 <- func(dict1)
dict3 <- func(dict2)

global dict: dict1


Fragment3:
output dict1

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java:243]: Why dict3 can ref dict2 before we use ScalarOperatorsReuseRule?

Comment (stdpain) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java:243]: but here we will add a decode node , may be it's not as expected.

**Comment by packy92** [comment]
run starrocks_admit_test

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**